File: Round 2/costume-change.cpp
// Copyright (c) 2018 kamyu. All rights reserved.

/*
 * Google Code Jam 2018 Round 2 - Problem C. Costume Change
 * https://codejam.withgoogle.com/2018/challenges/0000000000007706/dashboard/0000000000045875
 *
 * Time:  O(N^3 * sqrt(N))
 * Space: O(N)
 *
 */

#include <iostream>
#include <vector>
#include <queue>

using std::cin;
using std::cout;
using std::endl;
using std::vector;
using std::queue;

bool dfs(const vector<vector<int>>& edge,
         int u,
         const vector<int>& level,
         vector<bool> *vis,
         vector<int> *pair1,
         vector<int> *pair2) {
    (*vis)[u] = true;
    for (auto it = edge[u].begin(); it != edge[u].end(); ++it) {
        int v = (*pair2)[*it];
        if (v == -1 || (!(*vis)[v] && level[u] < level[v] &&
                        dfs(edge, v, level, vis, pair1, pair2))) {
            (*pair1)[u] = *it, (*pair2)[*it] = u;
            return true;
       }
    }
    return false;
}

int hopcroftKarp(const vector<vector<int>>& edge) {
    const int n = edge.size();  // n is the number of left nodes
    vector<bool> vis(n);
    vector<int> level(n);
    vector<int> pair1(n, -1);
    vector<int> pair2(n, -1);
    for (int match = 0; ;) {
        queue<int> Q;
        for (int i = 0; i < n; ++i) {
            if (pair1[i] == -1) {
                level[i] = 0;
                Q.emplace(i);
            } else {
                level[i] = -1;
            }
        }
        while (!Q.empty()) {
            int u = Q.front(); Q.pop();
            for (auto it = edge[u].begin(); it != edge[u].end(); ++it) {
                int v = pair2[*it];
                if (v != -1 && level[v] < 0) {
                    level[v] = level[u] + 1;
                    Q.emplace(v);
                }
            }
        }
        for (int i = 0; i < n; ++i) {
            vis[i] = false;
        }
        int d = 0;
        for (int i = 0; i < n; ++i) {
            if (pair1[i] == -1 && dfs(edge, i, level, &vis, &pair1, &pair2)) {
                ++d;
            }
        }
        if (d == 0) {
            return match;
        }
        match += d;
    }
}

int costume_change() {
    int N;
    cin >> N;
    vector<vector<int>> A(N, vector<int>(N));
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            cin >> A[i][j];
        }
    }
    int keep = 0;
    for (int color = -N; color <= N; ++color) {
        vector<vector<int>> edge(N);
        for (int i = 0; i < N; ++i) {
            for (int j = 0; j < N; ++j) {
                if (A[i][j] == color) {
                    edge[i].emplace_back(j);
                }
            }
        }
        keep += hopcroftKarp(edge);
    }
    return N * N - keep;
}

int main() {
    int T;
    cin >> T;
    for (int test = 1; test <= T; ++test) {
        cout << "Case #" << test << ": " << costume_change() << endl;
    }
    return 0;
}

File: Round 2/costume-change.py
# Copyright (c) 2018 kamyu. All rights reserved.
#
# Google Code Jam 2018 Round 2 - Problem C. Costume Change
# https://codingcompetitions.withgoogle.com/codejam/round/0000000000007706/0000000000045875
#
# Time:  O(sum(E * sqrt(V))) = O(N^2 * sqrt(N))
# Space: O(V) = O(N)
#

import collections

# Time:  O(E * sqrt(V))
# Space: O(V)
# Source code from http://code.activestate.com/recipes/123641-hopcroft-karp-bipartite-matching/
# Hopcroft-Karp bipartite max-cardinality matching and max independent set
# David Eppstein, UC Irvine, 27 Apr 2002
def bipartiteMatch(graph):
    '''Find maximum cardinality matching of a bipartite graph (U,V,E).
    The input format is a dictionary mapping members of U to a list
    of their neighbors in V.  The output is a triple (M,A,B) where M is a
    dictionary mapping members of V to their matches in U, A is the part
    of the maximum independent set in U, and B is the part of the MIS in V.
    The same object may occur in both U and V, and is treated as two
    distinct vertices if this happens.'''
    
    # initialize greedy matching (redundant, but faster than full search)
    matching = {}
    for u in graph:
        for v in graph[u]:
            if v not in matching:
                matching[v] = u
                break
    
    while 1:
        # structure residual graph into layers
        # pred[u] gives the neighbor in the previous layer for u in U
        # preds[v] gives a list of neighbors in the previous layer for v in V
        # unmatched gives a list of unmatched vertices in final layer of V,
        # and is also used as a flag value for pred[u] when u is in the first layer
        preds = {}
        unmatched = []
        pred = dict([(u,unmatched) for u in graph])
        for v in matching:
            del pred[matching[v]]
        layer = list(pred)
        
        # repeatedly extend layering structure by another pair of layers
        while layer and not unmatched:
            newLayer = {}
            for u in layer:
                for v in graph[u]:
                    if v not in preds:
                        newLayer.setdefault(v,[]).append(u)
            layer = []
            for v in newLayer:
                preds[v] = newLayer[v]
                if v in matching:
                    layer.append(matching[v])
                    pred[matching[v]] = v
                else:
                    unmatched.append(v)
        
        # did we finish layering without finding any alternating paths?
        if not unmatched:
            unlayered = {}
            for u in graph:
                for v in graph[u]:
                    if v not in preds:
                        unlayered[v] = None
            return (matching,list(pred),list(unlayered))

        # recursively search backward through layers to find alternating paths
        # recursion returns true if found path, false otherwise
        def recurse(v):
            if v in preds:
                L = preds[v]
                del preds[v]
                for u in L:
                    if u in pred:
                        pu = pred[u]
                        del pred[u]
                        if pu is unmatched or recurse(pu):
                            matching[v] = u
                            return 1
            return 0

        for v in unmatched: recurse(v)

def costume_change():
    N = input()
    A = []
    for _ in xrange(N):
        A.append(map(int, raw_input().strip().split()))
    keep = 0
    for color in xrange(-N, N+1):
        E = collections.defaultdict(list)
        for i in xrange(N):
            for j in xrange(N):
                if A[i][j] == color:
                    E[i].append(j)
        M, _, _ = bipartiteMatch(E)
        keep += len(M)
    return N**2 - keep

for case in xrange(input()):
    print 'Case #%d: %s' % (case+1, costume_change())
